File: cmake/CubismViewerBlindTest.cs
using System; using System.Collections.Generic; using System.Drawing;
using System.Drawing.Drawing2D; using System.Drawing.Imaging;
using System.Globalization; using System.IO; using System.Runtime.InteropServices;

public static class CubismViewerBlindTest
{
    const int Canvas = 800, Margin = 40;

    sealed class Frame { public string Name, Viewer, Native;
        public Bitmap ViewerImage, NativeImage; }

    sealed class Pixels : IDisposable
    {
        public Bitmap Image;
        public byte[] Data;
        public int Stride;
        public Pixels(Bitmap source)
        {
            Image = new Bitmap(source.Width, source.Height,
                PixelFormat.Format32bppArgb);
            using (Graphics g = Graphics.FromImage(Image))
                g.DrawImageUnscaled(source, 0, 0);
            BitmapData bits = Image.LockBits(new Rectangle(0, 0,
                Image.Width, Image.Height), ImageLockMode.ReadOnly,
                PixelFormat.Format32bppArgb);
            Stride = Math.Abs(bits.Stride);
            Data = new byte[Stride * Image.Height];
            Marshal.Copy(bits.Scan0, Data, 0, Data.Length);
            Image.UnlockBits(bits);
        }
        public int Offset(int x, int y) { return y * Stride + x * 4; }
        public void Dispose() { Image.Dispose(); }
    }

    struct Region { public string Name; public Rectangle Area;
        public Region(string name, Rectangle area) { Name = name; Area = area; }
    }

    static bool Visible(Color pixel, Color background, bool alpha)
    {
        if (alpha) return pixel.A > 8;
        int delta = Math.Max(Math.Abs(pixel.R - background.R),
            Math.Max(Math.Abs(pixel.G - background.G),
                Math.Abs(pixel.B - background.B)));
        return delta > 18;
    }

    static Rectangle Bounds(string path)
    {
        using (Bitmap image = new Bitmap(path))
        {
            Color background = image.GetPixel(image.Width / 2,
                Math.Min(image.Height - 1, Math.Max(1, image.Height / 20)));
            int alphaSamples = 0, nonOpaque = 0;
            for (int y = 0; y < image.Height; y += Math.Max(1, image.Height / 32))
            for (int x = 0; x < image.Width; x += Math.Max(1, image.Width / 32))
            {
                if (image.GetPixel(x, y).A < 250) nonOpaque++;
                alphaSamples++;
            }
            bool alpha = nonOpaque >= Math.Max(2, alphaSamples / 10);
            int size = image.Width * image.Height;
            bool[] foreground = new bool[size];
            for (int y = 0; y < image.Height; ++y)
            for (int x = 0; x < image.Width; ++x)
                foreground[y * image.Width + x] =
                    Visible(image.GetPixel(x, y), background, alpha);
            int[] queue = new int[size];
            int bestCount = 0, bestLeft = 0, bestTop = 0, bestRight = -1,
                bestBottom = -1;
            for (int start = 0; start < size; ++start)
            {
                if (!foreground[start]) continue;
                int head = 0, tail = 0, count = 0;
                int left = image.Width, top = image.Height, right = -1, bottom = -1;
                queue[tail++] = start; foreground[start] = false;
                while (head < tail)
                {
                    int current = queue[head++], x = current % image.Width;
                    int y = current / image.Width; count++;
                    left = Math.Min(left, x); top = Math.Min(top, y);
                    right = Math.Max(right, x); bottom = Math.Max(bottom, y);
                    int[] next = { current - 1, current + 1,
                        current - image.Width, current + image.Width };
                    for (int i = 0; i < next.Length; ++i)
                    {
                        int value = next[i];
                        if (value < 0 || value >= size || !foreground[value] ||
                            (i == 0 && x == 0) ||
                            (i == 1 && x == image.Width - 1)) continue;
                        foreground[value] = false; queue[tail++] = value;
                    }
                }
                if (count <= bestCount) continue;
                bestCount = count; bestLeft = left; bestTop = top;
                bestRight = right; bestBottom = bottom;
            }
            if (bestRight < bestLeft || bestBottom < bestTop)
                throw new InvalidOperationException("No model foreground in " + path);
            return Rectangle.FromLTRB(bestLeft, bestTop,
                bestRight + 1, bestBottom + 1);
        }
    }

    static Bitmap Normalize(string path, Rectangle anchor)
    {
        using (Bitmap source = new Bitmap(path))
        {
            double scale = Math.Min((Canvas - 2.0 * Margin) / anchor.Width,
                (Canvas - 2.0 * Margin) / anchor.Height);
            float left = (float)((Canvas - anchor.Width * scale) / 2.0 -
                anchor.Left * scale);
            float top = (float)((Canvas - anchor.Height * scale) / 2.0 -
                anchor.Top * scale);
            int padding = Math.Max(4, Math.Max(anchor.Width, anchor.Height) / 40);
            Rectangle crop = Rectangle.Intersect(new Rectangle(0, 0,
                source.Width, source.Height), new Rectangle(anchor.Left - padding,
                anchor.Top - padding, anchor.Width + padding * 2,
                anchor.Height + padding * 2));
            RectangleF destination = new RectangleF(
                left + crop.Left * (float)scale,
                top + crop.Top * (float)scale,
                crop.Width * (float)scale, crop.Height * (float)scale);
            Bitmap output = new Bitmap(Canvas, Canvas, PixelFormat.Format32bppArgb);
            using (Graphics g = Graphics.FromImage(output))
            {
                g.Clear(Color.White);
                g.CompositingMode = CompositingMode.SourceOver;
                g.InterpolationMode = InterpolationMode.HighQualityBicubic;
                g.PixelOffsetMode = PixelOffsetMode.HighQuality;
                g.DrawImage(source, destination, crop, GraphicsUnit.Pixel);
            }
            return output;
        }
    }

    static bool Foreground(Pixels pixels, int offset)
    {
        return Math.Min(pixels.Data[offset], Math.Min(pixels.Data[offset + 1],
            pixels.Data[offset + 2])) < 237;
    }

    static void Measure(TextWriter writer, Frame frame, Region region,
        Pixels viewer, Pixels native, Pixels viewerBase, Pixels nativeBase)
    {
        long count = 0, within = 0, intersection = 0, union = 0;
        double absolute = 0, viewerChange = 0, nativeChange = 0;
        for (int y = region.Area.Top; y < region.Area.Bottom; y += 2)
        for (int x = region.Area.Left; x < region.Area.Right; x += 2)
        {
            int a = viewer.Offset(x, y), b = native.Offset(x, y);
            int av = viewerBase.Offset(x, y), bn = nativeBase.Offset(x, y);
            bool first = Foreground(viewer, a), second = Foreground(native, b);
            if (first || second) union++;
            if (first && second) intersection++;
            if (!first && !second) continue;
            int maximum = 0;
            for (int channel = 0; channel < 3; ++channel)
            {
                int delta = Math.Abs(viewer.Data[a + channel] -
                    native.Data[b + channel]);
                maximum = Math.Max(maximum, delta);
                absolute += delta;
                viewerChange += Math.Abs(viewer.Data[a + channel] -
                    viewerBase.Data[av + channel]);
                nativeChange += Math.Abs(native.Data[b + channel] -
                    nativeBase.Data[bn + channel]);
            }
            if (maximum <= 8) within++;
            count++;
        }
        double similarity = count == 0 ? 100 :
            100 * (1 - absolute / (count * 3 * 255.0));
        double near = count == 0 ? 100 : 100.0 * within / count;
        double iou = union == 0 ? 100 : 100.0 * intersection / union;
        double vc = count == 0 ? 0 : viewerChange / (count * 3 * 255.0);
        double nc = count == 0 ? 0 : nativeChange / (count * 3 * 255.0);
        double ratio = vc < 0.000001 ? (nc < 0.000001 ? 1 : 999) : nc / vc;
        writer.WriteLine(String.Join(",", frame.Name, region.Name,
            similarity.ToString("F4", CultureInfo.InvariantCulture),
            near.ToString("F4", CultureInfo.InvariantCulture),
            iou.ToString("F4", CultureInfo.InvariantCulture),
            vc.ToString("F6", CultureInfo.InvariantCulture),
            nc.ToString("F6", CultureInfo.InvariantCulture),
            ratio.ToString("F4", CultureInfo.InvariantCulture), count));
    }

    static void Ballot(Frame frame, string path, bool viewerFirst)
    {
        using (Bitmap ballot = new Bitmap(Canvas * 2 + 30, Canvas + 60))
        using (Graphics g = Graphics.FromImage(ballot))
        using (Font font = new Font("Segoe UI", 20, FontStyle.Bold))
        {
            g.Clear(Color.FromArgb(238, 238, 238));
            Bitmap first = viewerFirst ? frame.ViewerImage : frame.NativeImage;
            Bitmap second = viewerFirst ? frame.NativeImage : frame.ViewerImage;
            g.DrawImageUnscaled(first, 0, 60);
            g.DrawImageUnscaled(second, Canvas + 30, 60);
            g.DrawString("A", font, Brushes.Black, Canvas / 2 - 10, 15);
            g.DrawString("B", font, Brushes.Black, Canvas + 30 + Canvas / 2 - 10, 15);
            ballot.Save(path, ImageFormat.Png);
        }
    }

    static string FindNative(string directory, string name)
    {
        foreach (string extension in new[] { ".bmp", ".png" })
        {
            string path = Path.Combine(directory, name + extension);
            if (File.Exists(path)) return path;
        }
        return null;
    }

    public static string Run(string viewerDirectory, string nativeDirectory,
        string outputDirectory, int seed)
    {
        Directory.CreateDirectory(outputDirectory);
        string normalized = Path.Combine(outputDirectory, "normalized");
        string ballots = Path.Combine(outputDirectory, "ballots");
        Directory.CreateDirectory(normalized); Directory.CreateDirectory(ballots);
        var frames = new List<Frame>();
        var viewerFiles = new List<string>();
        viewerFiles.AddRange(Directory.GetFiles(viewerDirectory, "*.png"));
        viewerFiles.AddRange(Directory.GetFiles(viewerDirectory, "*.bmp"));
        foreach (string viewer in viewerFiles)
        {
            string name = Path.GetFileNameWithoutExtension(viewer);
            string native = FindNative(nativeDirectory, name);
            if (native != null) frames.Add(new Frame { Name = name,
                Viewer = viewer, Native = native });
        }
        frames.Sort((a, b) => String.CompareOrdinal(a.Name, b.Name));
        if (frames.Count == 0) throw new InvalidOperationException("No frame pairs");
        Frame baseline = frames.Find(f => f.Name == "track-000") ?? frames[0];
        string cleanViewer = Path.Combine(viewerDirectory, "idle.png");
        Rectangle viewerBounds = Bounds(File.Exists(cleanViewer)
            ? cleanViewer : baseline.Viewer);
        Rectangle nativeBounds = Bounds(baseline.Native);
        foreach (Frame frame in frames)
        {
            frame.ViewerImage = Normalize(frame.Viewer, viewerBounds);
            frame.NativeImage = Normalize(frame.Native, nativeBounds);
            CubismViewerImageMask.KeepLargest(frame.ViewerImage);
            CubismViewerImageMask.KeepLargest(frame.NativeImage);
            frame.ViewerImage.Save(Path.Combine(normalized,
                frame.Name + "-viewer.png"), ImageFormat.Png);
            frame.NativeImage.Save(Path.Combine(normalized,
                frame.Name + "-native.png"), ImageFormat.Png);
        }
        if (seed == 0) seed = Environment.TickCount & Int32.MaxValue;
        Random random = new Random(seed);
        Region[] regions = {
            new Region("full", new Rectangle(0, 0, Canvas, Canvas)),
            new Region("face", new Rectangle(225, 170, 350, 285)),
            new Region("hair", new Rectangle(70, 40, 660, 455)),
            new Region("hands", new Rectangle(55, 395, 690, 315)) };
        string metricsPath = Path.Combine(outputDirectory, "metrics.csv");
        string keyPath = Path.Combine(outputDirectory, "answer-key.csv");
        using (var metrics = new StreamWriter(metricsPath, false))
        using (var key = new StreamWriter(keyPath, false))
        using (var vb = new Pixels(baseline.ViewerImage))
        using (var nb = new Pixels(baseline.NativeImage))
        {
            metrics.WriteLine("frame,region,similarity,within8,foreground_iou," +
                "viewer_change,native_change,response_ratio,samples");
            key.WriteLine("frame,A,B");
            foreach (Frame frame in frames)
            using (var v = new Pixels(frame.ViewerImage))
            using (var n = new Pixels(frame.NativeImage))
            {
                foreach (Region region in regions)
                    Measure(metrics, frame, region, v, n, vb, nb);
                bool viewerFirst = random.Next(2) == 0;
                Ballot(frame, Path.Combine(ballots, frame.Name + ".png"), viewerFirst);
                key.WriteLine(frame.Name + "," +
                    (viewerFirst ? "Viewer,Native" : "Native,Viewer"));
            }
        }
        File.WriteAllText(Path.Combine(outputDirectory, "summary.json"),
            "{\n  \"seed\": " + seed + ",\n  \"frames\": " + frames.Count +
            ",\n  \"canvas\": " + Canvas + "\n}\n");
        foreach (Frame frame in frames) { frame.ViewerImage.Dispose();
            frame.NativeImage.Dispose(); }
        return outputDirectory;
    }
}

File: src/tools/cubism_viewer_blind_test.c
#define WIN32_LEAN_AND_MEAN
#include <windows.h>
#include <objbase.h>

#include "cubism_viewer_blind_frames.h"
#include "validation_image.h"
#include "windows_tool.h"

#include <limits.h>
#include <locale.h>
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <wchar.h>

enum { CANVAS = 800, MARGIN = 40 };

typedef struct Region {
    const char *name;
    int left, top, right, bottom;
} Region;

typedef struct DotNetRandom {
    int seed[56];
    int inext;
    int inextp;
} DotNetRandom;

static int foreground_pixel(const unsigned char *pixel) {
    return min(pixel[0], min(pixel[1], pixel[2])) < 237;
}

static int write_metrics(FILE *writer, const Frame *frame, const Region *region,
    const BongoCatValidationImage *viewer, const BongoCatValidationImage *native,
    const BongoCatValidationImage *viewer_base,
    const BongoCatValidationImage *native_base) {
    long long count = 0, within = 0, intersection = 0, union_count = 0;
    double absolute = 0, viewer_change = 0, native_change = 0;
    int x, y;
    for (y = region->top; y < region->bottom; y += 2)
    for (x = region->left; x < region->right; x += 2) {
        const unsigned char *a = viewer->bgra + ((size_t)y * viewer->width + x) * 4;
        const unsigned char *b = native->bgra + ((size_t)y * native->width + x) * 4;
        const unsigned char *av = viewer_base->bgra +
            ((size_t)y * viewer_base->width + x) * 4;
        const unsigned char *bn = native_base->bgra +
            ((size_t)y * native_base->width + x) * 4;
        int first = foreground_pixel(a), second = foreground_pixel(b);
        int maximum = 0, channel;
        if (first || second) ++union_count;
        if (first && second) ++intersection;
        if (!first && !second) continue;
        for (channel = 0; channel < 3; ++channel) {
            int delta = abs((int)a[channel] - b[channel]);
            maximum = max(maximum, delta); absolute += delta;
            viewer_change += abs((int)a[channel] - av[channel]);
            native_change += abs((int)b[channel] - bn[channel]);
        }
        if (maximum <= 8) ++within;
        ++count;
    }
    {
        double similarity = count == 0 ? 100.0 : 100.0 *
            (1.0 - absolute / (count * 3.0 * 255.0));
        double near_value = count == 0 ? 100.0 : 100.0 * within / count;
        double iou = union_count == 0 ? 100.0 : 100.0 * intersection / union_count;
        double viewer_ratio = count == 0 ? 0.0 : viewer_change /
            (count * 3.0 * 255.0);
        double native_ratio = count == 0 ? 0.0 : native_change /
            (count * 3.0 * 255.0);
        double ratio = viewer_ratio < 0.000001 ?
            (native_ratio < 0.000001 ? 1.0 : 999.0) : native_ratio / viewer_ratio;
        return fprintf(writer, "%ls,%s,%.4f,%.4f,%.4f,%.6f,%.6f,%.4f,%lld\n",
            frame->name, region->name, similarity, near_value, iou, viewer_ratio,
            native_ratio, ratio, count) >= 0;
    }
}

static int write_ballot(const wchar_t *path, const Frame *frame, int viewer_first) {
    BongoCatValidationImage ballot = {0};
    const BongoCatValidationImage *first = viewer_first ? &frame->viewer : &frame->native;
    const BongoCatValidationImage *second = viewer_first ? &frame->native : &frame->viewer;
    int y, x;
    ballot.width = CANVAS * 2 + 30; ballot.height = CANVAS + 60;
    ballot.bgra = (unsigned char *)malloc((size_t)ballot.width * ballot.height * 4);
    if (!ballot.bgra) return 0;
    for (y = 0; y < ballot.height; ++y) for (x = 0; x < ballot.width; ++x) {
        unsigned char *pixel = ballot.bgra + ((size_t)y * ballot.width + x) * 4;
        pixel[0] = pixel[1] = pixel[2] = 238; pixel[3] = 255;
    }
    for (y = 0; y < CANVAS; ++y) {
        memcpy(ballot.bgra + ((size_t)(y + 60) * ballot.width) * 4,
            first->bgra + (size_t)y * CANVAS * 4, CANVAS * 4);
        memcpy(ballot.bgra + ((size_t)(y + 60) * ballot.width + CANVAS + 30) * 4,
            second->bgra + (size_t)y * CANVAS * 4, CANVAS * 4);
    }
    if (!bongo_cat_validation_image_draw_label(&ballot, L"A",
            CANVAS / 2.0f - 10.0f, 15.0f) ||
        !bongo_cat_validation_image_draw_label(&ballot, L"B",
            CANVAS + 30.0f + CANVAS / 2.0f - 10.0f, 15.0f)) {
        bongo_cat_validation_image_free(&ballot); return 0;
    }
    x = bongo_cat_validation_image_save(path, &ballot, 1);
    bongo_cat_validation_image_free(&ballot);
    return x;
}

static void random_init(DotNetRandom *random, int seed) {
    int subtraction = seed == INT_MIN ? INT_MAX : abs(seed);
    int mj = 161803398 - subtraction, mk = 1, i, ii, k;
    memset(random, 0, sizeof(*random));
    random->seed[55] = mj;
    for (i = 1; i < 55; ++i) {
        ii = (21 * i) % 55; random->seed[ii] = mk;
        mk = mj - mk; if (mk < 0) mk += INT_MAX; mj = random->seed[ii];
    }
    for (k = 1; k < 5; ++k) for (i = 1; i < 56; ++i) {
        int n = 1 + (i + 30) % 55;
        random->seed[i] -= random->seed[n];
        if (random->seed[i] < 0) random->seed[i] += INT_MAX;
    }
    random->inext = 0; random->inextp = 21;
}

static int random_next_two(DotNetRandom *random) {
    int loc, locp, value;
    loc = random->inext + 1; if (loc >= 56) loc = 1; random->inext = loc;
    locp = random->inextp + 1; if (locp >= 56) locp = 1; random->inextp = locp;
    value = random->seed[loc] - random->seed[locp];
    if (value == INT_MAX) --value;
    if (value < 0) value += INT_MAX;
    random->seed[loc] = value;
    return (int)((value / (double)INT_MAX) * 2.0);
}

static int write_summary(const wchar_t *path, int seed, size_t frames) {
    FILE *file = _wfopen(path, L"wb");
    if (!file) return 0;
    fprintf(file, "{\n  \"seed\": %d,\n  \"frames\": %zu,\n  \"canvas\": 800\n}\n",
        seed, frames);
    return fclose(file) == 0;
}

int wmain(int argc, wchar_t **argv) {
    FrameList frames = {0};
    wchar_t *normalized = NULL, *ballots = NULL, *metrics_path = NULL;
    wchar_t *key_path = NULL, *summary_path = NULL;
    BongoCatValidationRect viewer_bounds, native_bounds;
    size_t baseline_index = 0, index;
    int seed, result = 1;
    const Region regions[] = {
        {"full", 0, 0, CANVAS, CANVAS},
        {"face", 225, 170, 575, 455},
        {"hair", 70, 40, 730, 495},
        {"hands", 55, 395, 745, 710}
    };
    setlocale(LC_NUMERIC, "C");
    if (argc < 4) {
        fwprintf(stderr, L"usage: blind-test.exe viewer native output [seed]\n");
        return 2;
    }
    seed = argc > 4 ? _wtoi(argv[4]) : 20260807;
    if (seed == 0) seed = (int)(GetTickCount() & INT_MAX);
    if (FAILED(CoInitializeEx(NULL, COINIT_MULTITHREADED))) return 1;
    if (!bongo_cat_tool_ensure_directory(argv[3]) ||
        !blind_collect_frames(argv[1], argv[2], &frames))
        goto done;
    normalized = blind_join_path(argv[3], L"normalized");
    ballots = blind_join_path(argv[3], L"ballots");
    metrics_path = blind_join_path(argv[3], L"metrics.csv");
    key_path = blind_join_path(argv[3], L"answer-key.csv");
    summary_path = blind_join_path(argv[3], L"summary.json");
    if (!normalized || !ballots || !metrics_path || !key_path || !summary_path ||
        !bongo_cat_tool_ensure_directory(normalized) ||
        !bongo_cat_tool_ensure_directory(ballots)) goto done;
    for (index = 0; index < frames.count; ++index)
        if (wcscmp(frames.items[index].name, L"track-000") == 0) baseline_index = index;
    {
        wchar_t *idle = blind_join_suffix(argv[1], L"idle", L".png");
        int have_idle = idle && _waccess(idle, 0) == 0;
        wchar_t *viewer_bounds_path = have_idle ? idle : frames.items[baseline_index].viewer_path;
        BongoCatValidationImage bounds_image = {0};
        if (!bongo_cat_validation_image_load(viewer_bounds_path, &bounds_image) ||
            !bongo_cat_validation_image_bounds(&bounds_image, &viewer_bounds)) {
            bongo_cat_validation_image_free(&bounds_image); free(idle); goto done;
        }
        bongo_cat_validation_image_free(&bounds_image);
        if (!bongo_cat_validation_image_load(frames.items[baseline_index].native_path,
                &bounds_image) || !bongo_cat_validation_image_bounds(&bounds_image,
                &native_bounds)) {
            bongo_cat_validation_image_free(&bounds_image); free(idle); goto done;
        }
        bongo_cat_validation_image_free(&bounds_image); free(idle);
    }
    for (index = 0; index < frames.count; ++index) {
        wchar_t *viewer_output, *native_output;
        if (!bongo_cat_validation_image_normalize_file(
                frames.items[index].viewer_path, viewer_bounds,
                &frames.items[index].viewer) ||
            !bongo_cat_validation_image_normalize_file(
                frames.items[index].native_path, native_bounds,
                &frames.items[index].native) ||
            !bongo_cat_validation_image_keep_largest(&frames.items[index].viewer) ||
            !bongo_cat_validation_image_keep_largest(&frames.items[index].native)) {
            goto done;
        }
        viewer_output = blind_join_suffix(normalized, frames.items[index].name, L"-viewer.png");
        native_output = blind_join_suffix(normalized, frames.items[index].name, L"-native.png");
        if (!viewer_output || !native_output ||
            !bongo_cat_validation_image_save(viewer_output, &frames.items[index].viewer, 1) ||
            !bongo_cat_validation_image_save(native_output, &frames.items[index].native, 1)) {
            free(viewer_output); free(native_output); goto done;
        }
        free(viewer_output); free(native_output);
    }
    {
        FILE *metrics = _wfopen(metrics_path, L"w");
        FILE *key = _wfopen(key_path, L"w");
        DotNetRandom random;
        if (!metrics || !key) { if (metrics) fclose(metrics); if (key) fclose(key); goto done; }
        fprintf(metrics, "frame,region,similarity,within8,foreground_iou,viewer_change,native_change,response_ratio,samples\n");
        fprintf(key, "frame,A,B\n");
        random_init(&random, seed);
        for (index = 0; index < frames.count; ++index) {
            size_t region;
            for (region = 0; region < sizeof(regions) / sizeof(regions[0]); ++region)
                if (!write_metrics(metrics, &frames.items[index], &regions[region],
                    &frames.items[index].viewer, &frames.items[index].native,
                    &frames.items[baseline_index].viewer, &frames.items[baseline_index].native)) {
                    fclose(metrics); fclose(key); goto done;
                }
            {
                int viewer_first = random_next_two(&random) == 0;
                wchar_t *ballot = blind_join_suffix(ballots, frames.items[index].name, L".png");
                if (!ballot || !write_ballot(ballot, &frames.items[index], viewer_first)) {
                    free(ballot); fclose(metrics); fclose(key); goto done;
                }
                free(ballot);
                fprintf(key, "%ls,%s\n", frames.items[index].name,
                    viewer_first ? "Viewer,Native" : "Native,Viewer");
            }
        }
        fclose(metrics); fclose(key);
    }
    if (!write_summary(summary_path, seed, frames.count)) goto done;
    wprintf(L"%ls\n", argv[3]);
    result = 0;
done:
    free(normalized); free(ballots); free(metrics_path); free(key_path); free(summary_path);
    blind_free_frames(&frames);
    CoUninitialize();
    if (result) fwprintf(stderr, L"blind test failed\n");
    return result;
}
